### PR TITLE
KAFKA-17209: Revisit testCurrentLag for asyncConsumer

### PR DESCRIPTION
**Pull request by TaiJuWu**
jira: https://issues.apache.org/jira/browse/KAFKA-17209
## More detailed description of your change
The difference between `ClassicConsumer` and `AsyncConsumer` is ClassicConsumer is single thread and AsyncConsumer is two thread, one is app thread, the other is background thread.

`Fetch` Request is generated by background thread and `ListOffset` Request is generated by application thread.
This will make `testCurrentLag` fail if AsyncConsumer generated `Fetch Request`  first.
So we support out of order feature for `MockClient` to resolve issue.

Another difference between asyncConsumer and ClassicConsumer is front do generate additional `FetchOffset` Request from CommitRequestManager but ClassicConsumer does not have the request.

The reason is  coordinator is not ready during test for `ClassicConsumer`  so we can ignore this request for `AsyncConsumer`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 1]
```diff
@@ -2491,7 +2508,8 @@ public void testCurrentLag(GroupProtocol groupProtocol) {
         client.respond(listOffsetsResponse(singletonMap(tp0, 90L)));
         consumer.poll(Duration.ofMillis(0));
 
-        assertEquals(OptionalLong.of(40L), consumer.currentLag(tp0));
+        TestUtils.waitForCondition(() -> OptionalLong.of(40L).equals(consumer.currentLag(tp0)),
+                "Consumer can't get excepted lag.");
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: While this change may allow the test to pass, it suggests to me that the new consumer is not keeping the semantics for the `currentLag()` API, in that it should return the correct value _immediately_, not _eventually_.

**Review by kirktrue — COMMENTED** [review]
Thanks for the PR, @TaiJuWu!

My impression is that these changes relax the requirements in order to allow the new consumer to pass successfully. I'd like to avoid `if` statements at the individual test layer as it could lead to “wallpapering” over problems at the implementation layer.

Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 1 above)
Comment (TaiJuWu) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: > While this change may allow the test to pass, it suggests to me that the new consumer is not keeping the semantics for the `currentLag()` API, in that it should return the correct value _immediately_, not _eventually_.

Hi @kirktrue ,thanks for review.
I think this condition can be removed, let's test with CI.

**Comment by TaiJuWu** [comment]
> Thanks for the PR, @TaiJuWu!
> 
> My impression is that these changes relax the requirements in order to allow the new consumer to pass successfully. I'd like to avoid `if` statements at the individual test layer as it could lead to “wallpapering” over problems at the implementation layer.
> 
> Thanks!


Hi @kirktrue , thanks for your review and comments.
I removed `if` statement and rewrite `waitForUntil` condition, let's test on CI.

**Pull request closed without merging**